### ISSUE TITLE
Organize modules for 1.0.0

### ISSUE THREAD

**Issue by jacob-alford** [jacob-alford/schemata-ts#166]
For 1.0.0 I'd like the modules to follow this structure (all in src):

1. base (contains all the base instances, SchemableBase, Schemable2 currently)
2. internal (contains everything in `internal` that isn't in base)
3. schemables (all the With typeclasses, see #167)
4. schemata (same as it is now)
5. `Arbitrary.ts`
6. `Encoder.ts`
7. `Eq.ts`
8. `Guard.ts`
9. `PatternBuilder.ts`
10. `SchemaExt.ts`
11. `SchemableExt.ts`
12. `TaskDecoder.ts`
13. `Type.ts`
14. `index.ts` (contains qualified exports from schemata.ts as `schemata`, and interpreters.ts as `interpreters`)
15. `interpreters.ts` (all partially applied named exports for all the schemable instances besides Schema)
16. `schemata.ts` (generated file that consists of all named exports from `schemata` subdirectories, and all named exports from `base / SchemaBase.ts` (see #168)

**Comment by skeate** [comment]
one thought re: architecture. When should a thing be a `WithFoo` interface and when should it just be a `schemable`?

E.g. `int` and `float` are in `src/number` right now, but they could also be done as `WithInt` and `WithFloat` (or `WithRangedNumbers` or something if you want to combine them)

What criteria should be used to distinguish which style is preferred? I have a vague feeling but nothing super concrete.

**Comment by jacob-alford** [comment]
Depends on #109, and #126

**Comment by jacob-alford** [comment]
That's a good question, because they're basically the same thing: properties on `SchemableExt`.  I think a good distinction would be if they ought to have definitions in the `Base` files or not.  To me, `Int` / `Float` / etc are nice to have in their own file because you might be more concerned about the internal behavior of those.  And so locating that behavior is easier that way – especially if I were a user of the library. But for more abstract types (like Invariant / Pattern / Brand) it makes more sense to be in the Base files.

**Comment by jacob-alford** [comment]
And another thing we could do in move all the `With` definitions to `Schemable2` to mirror io-ts, but I kind of like them in their own file

**Comment by jacob-alford** [comment]
Although now that you mention it, it'd be pretty nice to add a few more `With`s.  Just spit balling:

- `WithRangedNumbers`
- `WithFromString` or `WithParse`? (could house all the `FromString`s
- `WithDate`

One upside would be not needing to rely on the `generate:schemables` script for the JsDoc comments above the definition.  The downside would be that instance behavior is not co-located.

**Comment by jacob-alford** [comment]
The other option would be to meet in the middle and change the SchemableParams to WithModule, and just make them all their own With, like WithInt, WithFloat, etc, and stick them in the schemables directory.  

And this also brings up the possibility of specializing Schema to a `subset` of SchemableExt which could allow for more specialized behavior.  Though I think restricting schemas to behavior that works for every instance would be preferable, so maybe it doesn't matter anyway

**Comment by jacob-alford** [comment]
Okay, I've rambled a lot, but now I'm leaning toward a ticket that would be named something like this:

Replace schemables with their own `With` typeclass (consolidating where possible), update the generator script, and place all modules in the `schemables` directory.  

I think this would be a good consolidation / simplification.  And all of the instances (decoder/etc) would live in the same file.

What do you think @skeate?

**Comment by jacob-alford** [comment]
https://github.com/jacob-alford/schemata-ts/issues/167
